### DISCUSSION TITLE
Getting stimuli based on participant ID

### DISCUSSION THREAD

**Discussion by axia90** [jspsych/jsPsych#2744] [General]
Hi,

I have a CSV file containing all the trials I'd like to present to a participant, based on their ID. Here is the code I have at the beginning of the experiment, which prompts participants to put in their ID:

```js
const jsPsych = initJsPsych({
    timeline: [participant_id],
    experiment_width: 750,
    on_finish: function () {
        var id = jsPsych.data.get().filter({trial_purpose: 'id'}).values()[0].response.id;
        jsPsych.data.addProperties({sub_id: id});
        var file_name = 'sub_' + id + '_data.csv';
        jsPsych.data.displayData()
        jsPsych.data.get().localSave('csv', file_name)
    }
})
```

After that, I tried to filter all my trials using that ID number, like so:
```js
var session_trials = $.csv.toObjects(trials_csv).filter(
        x => x.sub_id == parseInt(jsPsych.data.get().filter({trial_purpose: 'id'}).values()[0].response.id)
)
```

However, the error I keep getting is: Cannot read properties of undefined (reading 'response')

I suspect this is something about the order of what's being run, but I'm not sure. Any help would be appreciated! Thanks!

**Comment by becky-gilbert** [comment]
I'm not sure without seeing the rest of your code. Did you put the `session_trials` line in your main script? Or inside the experiment's `on_finish` function, i.e.
```js
on_finish: function () {
  var id = jsPsych.data.get().filter({trial_purpose: 'id'}).values()[0].response.id;
  jsPsych.data.addProperties({sub_id: id});
  var file_name = 'sub_' + id + '_data.csv';
  jsPsych.data.displayData()
  jsPsych.data.get().localSave('csv', file_name)
  var session_trials = $.csv.toObjects(trials_csv)
    .filter(x => x.sub_id == parseInt(jsPsych.data.get().filter({trial_purpose: 'id'}).values()[0].response.id))
}
```
What I would do is add some `console.log`s to see what the values are at each step in the code you're using to retrieve the ID, especially the point right before you try to access `response`, since that seems to be causing the problem. E.g.
```js
on_finish: function() {
  // ...
  console.log("ID trial data: ", jsPsych.data.get().filter({trial_purpose: 'id'}).values()[0])
}
```
Looking at this output should give you some clues as to why `response` is undefined.
Another suggestion is to break down the `session_trials` line into different parts/variables. For instance, first create a variable to store the ID, then do the other stuff. 
I hope that helps!
